### PR TITLE
[fix](nereids)fix bug of unnest subuqery with having clause

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

1. run PushDownFilterThroughAggregation, PushDownFilterThroughProject and MergeFilters before subquery unnesting
2. should keep plan unchanged even left semi join's condition is true (because the right table may be empty())
3. PushDownFilterThroughProject need check if filter's input slots are all coming from project's output

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Review by 924060929 — DISMISSED** [review]
nice fix

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownFilterThroughProject.java [hunk 1]
```diff
@@ -53,27 +60,56 @@ public List<Rule> buildRules() {
...
 
     /** pushdown Filter through project */
-    public static Plan pushdownFilterThroughProject(LogicalFilter<LogicalProject<Plan>> filter) {
+    private static Plan pushdownFilterThroughProject(LogicalFilter<LogicalProject<Plan>> filter) {
         LogicalProject<Plan> project = filter.child();
-        return project.withChildren(
+        Set<Slot> childOutputs = project.getOutputSet();
+        Pair<Set<Expression>, Set<Expression>> splitConjuncts =
+                splitConjunctsByChildOutput(filter.getConjuncts(), childOutputs);
+        if (splitConjuncts.second.isEmpty()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownFilterThroughProject.java:74]: add some comment to explain why need this if statement

**Review by morrySnow — APPROVED** [review]

**Review by 924060929 — APPROVED** [review]

**Pull request merged**